rules/prefer-optional-catch-binding.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+const getDocumentationUrl = require('./utils/get-documentation-url');
+const {findVariable, isOpeningParenToken, isClosingParenToken} = require('eslint-utils');
+
+const ERROR_MESSAGE_ID = 'error';
+
+const selector = [
+	'CatchClause',
+	'>',
+	'Identifier.param'
+].join('');
+
+const create = context => {
+	return {
+		[selector]: node => {
+			const scope = context.getScope();
+			const variable = findVariable(scope, node);
+
+			if (variable.references.length !== 0) {
+				return;
+			}
+
+			const {name} = node;
+
+			context.report({
+				node,
+				messageId: ERROR_MESSAGE_ID,
+				data: {name},
+				fix: fixer => {
+					const tokenBefore = context.getTokenBefore(node);
+					const tokenAfter = context.getTokenAfter(node);
+
+					/* istanbul ignore next */
+					if (!isOpeningParenToken(tokenBefore) || !isClosingParenToken(tokenAfter)) {
+						throw new Error('Unexpected token.');
+					}
```

Comment:
I don't want fixer remove the wrong token, not sure is there a case can make this throws.

Reply:
How is the stack trace in ESLint if you throw? It's important that the user can see where it failed, so we can fix.

If the stack trace is not useful, maybe do a `context.report` here instead?

Reply:
How is the stack trace in ESLint if you throw? It's important that the user can see where it failed, so we can fix.

If the stack trace is not useful, maybe do a `context.report` here instead?

Reply:
Eslint will crash, with this

```text
\eslint-plugin-unicorn\node_modules\eslint\lib\linter\linter.js:1181
            throw err;
            ^

Error: Unexpected token.
Occurred while linting \eslint-plugin-unicorn\1.js:1
    at Object.fix (\eslint-plugin-unicorn\rules\prefer-optional-catch-binding.js:35:13)
    at normalizeFixes (\eslint-plugin-unicorn\node_modules\eslint\lib\linter\report-translator.js:178:28)
    at \eslint-plugin-unicorn\node_modules\eslint\lib\linter\report-translator.js:343:49
    at Object.report (\eslint-plugin-unicorn\node_modules\eslint\lib\linter\linter.js:905:41)
    at CatchClause>Identifier.param (\eslint-plugin-unicorn\rules\prefer-optional-catch-binding.js:25:12)
    at \eslint-plugin-unicorn\node_modules\eslint\lib\linter\safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (\eslint-plugin-unicorn\node_modules\eslint\lib\linter\safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (\eslint-plugin-unicorn\node_modules\eslint\lib\linter\node-event-generator.js:254:26)
    at NodeEventGenerator.applySelectors (\eslint-plugin-unicorn\node_modules\eslint\lib\linter\node-event-generator.js:283:22) {
  currentNode: Node {
    type: 'Identifier',
    start: 13,
    end: 14,
    loc: SourceLocation { start: [Position], end: [Position] },
    range: [ 13, 14 ],
    name: 'e',
    parent: Node {
      type: 'CatchClause',
      start: 7,
      end: 17,
      loc: [SourceLocation],
      range: [Array],
      param: [Circular],
      body: [Node],
      parent: [Node]
    }
  }
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

like `SyntaxError` wrong fixing 

```text
E:\project\@forks\eslint-plugin-unicorn\1.js
  1:13  error  Parsing error: Unexpected token )

✖ 1 problem (1 error, 0 warnings)
```

But more detail.